datafusion/functions/src/string/lower.rs

```diff
@@ -62,6 +62,6 @@ impl ScalarUDFImpl for LowerFunc {
     }
 
     fn invoke(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
-        handle(args, |string| string.to_lowercase(), "lower")
+        case_conversion(args, |string| string.to_lowercase(), "lower")
```

Comment:
It seems like to_lowercase actually works on unicode values, so it may change the length of the strings: https://doc.rust-lang.org/std/primitive.str.html#method.to_lowercase

Reply:
Thank you @alamb  for pointing out the mistake. In order to achieve this optimization, do you agree to implement the `to_lower` and `to_upper` functions by emulating the str's `to_lowercase` and `to_uppercase` functions, respectively?

Reply:
Maybe we can take some inspiration from how `to_uppercase` is implemented:

https://doc.rust-lang.org/src/alloc/str.rs.html#368

It seems like it should take advantage of the case case where the string is entirely ASCII

Perhaps you can use  `StringArrayBuilder` added in https://github.com/apache/arrow-datafusion/pull/9732, to copy data, adjusting the offsets if necessary 🤔

Reply:
Update -- I see you have already started on this -- I will try and take a more thorough look at this PR tomorrow;